packages/@sanity/form-builder/src/inputs/ImageInput/Asset.js

```diff
@@ -0,0 +1,254 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import Spinner from 'part:@sanity/components/loading/spinner'
+import DropDownButton from 'part:@sanity/components/buttons/dropdown'
+import Dialog from 'part:@sanity/components/dialogs/default'
+import DialogContent from 'part:@sanity/components/dialogs/content'
+import {List, Item} from 'part:@sanity/components/lists/default'
+import {WithReferringDocuments} from 'part:@sanity/base/with-referring-documents'
+import Preview from '../../Preview'
+import schema from 'part:@sanity/base/schema'
+import {IntentLink} from 'part:@sanity/base/router'
+import client from 'part:@sanity/base/client'
+
+import styles from './styles/Asset.css'
+import TrashIcon from 'react-icons/lib/md/delete'
+import LinkIcon from 'part:@sanity/base/link-icon'
+import MoreVertIcon from 'part:@sanity/base/more-vert-icon'
+
+import {get, startsWith} from 'lodash'
+
+export default class Asset extends React.PureComponent {
+  static propTypes = {
+    asset: PropTypes.shape({
+      _id: PropTypes.string,
+      referenceCount: PropTypes.number,
+      url: PropTypes.string
+    }),
+    onClick: PropTypes.func,
+    onKeyPress: PropTypes.func,
+    onDeleteComplete: PropTypes.func
+  }
+
+  state = {
+    showDialog: false,
+    isDeleting: false,
+    dialogType: undefined
+  }
+
+  handleDeleteAsset = asset => {
+    const {onDeleteComplete} = this.props
+    this.setState({isDeleting: true})
+
+    return client
+      .delete(asset._id)
+      .then(() => {
+        this.setState({
+          showDialog: false,
+          isDeleting: false
+        })
+        onDeleteComplete()
+      })
+      .catch(err => {
+        this.setState({
+          showDialog: true,
+          isDeleting: false,
+          dialogType: 'cantDelete'
```

Comment:
I'd prefer another name for this. As it stands now, this is a general error handling and more like "Delete failed for some reason". I think it's possible to distinguish between an error due to existing references and other kind of failures.

Reply:
This is outdated